[clang][FMV][AArch64] Improve streaming mode compatibility.

labrinea opened the pull request:
* Allow arm-streaming if all the functions versions adhere to it.
* Allow arm-streaming-compatible if all the functions versions adhere to it.

When the caller needs to toggle the streaming mode all the function versions of the callee must adhere to the same mode, otherwise the call will yield a runtime error.

Imagine the versions of the callee live in separate TUs. The version that is visible to the caller will determine the calling convention used when generating code for the callsite. Therefore we cannot support mixing streaming with non-streaming function versions. Imagine TU1 has a streaming caller and calls foo._sme which is streaming-compatible. The codegen for the callsite will not switch off the streaming mode. Then in TU2 we have a version which is non-streaming and could potentially be called in streaming mode. Similarly if the caller is non-streaming and the called version is streaming-compatible the codegen for the callsite will not switch on the streaming mode, but other versions may be streaming.

labrinea committed:
[clang][FMV][AArch64] Improve streaming mode compatibility.

* Allow arm-streaming if all the functions versions adhere to it.
* Allow arm-streaming-compatible if all the functions versions adhere to it.
* Allow arm-locally-streaming regardless of the other functions versions.

When the caller needs to toggle the streaming mode all the function versions
of the callee must adhere to the same mode, otherwise the call will yield a
runtime error.

Imagine the versions of the callee live in separate TUs. The version that
is visible to the caller will determine the calling convention used when
generating code for the callsite. Therefore we cannot support mixing
streaming with non-streaming function versions. Imagine TU1 has a streaming
caller and calls foo._sme which is streaming-compatible. The codegen for
the callsite will not switch off the streaming mode. Then in TU2 we have
a version which is non-streaming and could potentially be called in
streaming mode. Similarly if the caller is non-streaming and the called
version is streaming-compatible the codegen for the callsite will not
switch on the streaming mode, but other versions may be streaming.

jroelofs commented on clang/include/clang/AST/ASTContext.h:
```diff
@@ -3183,6 +3183,22 @@ class ASTContext : public RefCountedBase<ASTContext> {
+      if (!hasSameType(FD1->getParamDecl(I)->getOriginalType(),
+                       FD2->getParamDecl(I)->getOriginalType()))
+        return false;
+
```
```
if (FD1->isVariadic() != FD2->isVariadic())
  return false;
```

jroelofs reviewed (approved):
As a follow-up, I think we should consider making `__init_cpu_features_resolver()` a streaming-compatible function.

jroelofs commented:
relevant issue I've been meaning to file: https://github.com/llvm/llvm-project/issues/100200

labrinea committed:
Changes from last revision:

* Disregard declarations with different variadic type. Note that
  we are not diagnosing such differences, we just do not consider
  the two declarations part of the same declaration chain. As a
  result the diagnostic comes upon use: "ambiguous call". This
  is NFC.

* Added a sema test for variadic type mismatch.

* Added a codegen test for the calling conventions.

sdesmalen-arm commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -11145,7 +11148,24 @@ bool Sema::areMultiversionVariantFunctionsCompatible(
+
+    bool ArmStreamingCCMismatched = false;
+    // Locally streaming does not affect the calling convention.
+    if (OldFPT && NewFPT && !OldFD->hasAttr<ArmLocallyStreamingAttr>() &&
+        !NewFD->hasAttr<ArmLocallyStreamingAttr>()) {
```
Why is it necessary to check for `ArmLocallyStreamingAttr` here?

labrinea replied:
Because arm_locally_streaming is compatible with everything else, therefore if one of the two declarations has it, then we should not check whether the other declaration has arm_streaming or arm_streaming_compatible.

labrinea replied:
Are you saying that `ok_arm_streaming` and `ok_arm_streaming_compatible` from my example in `clang/test/Sema/aarch64-fmv-streaming.c` should report an error? If that's not what you meant, then we must not remove this check. Doing so will diagnose a mismatch because __arm_locally_streaming does not set neither of `FunctionType::SME_PStateSMEnabledMask` or `FunctionType::SME_PStateSMCompatibleMask`.

labrinea replied:
Perhaps I am mistaken. Do we need to add arm_streaming/arm_streaming_compatible to an arm_locally_streaming callee if the other versions have it?

sdesmalen-arm replied:
> Are you saying that ok_arm_streaming and ok_arm_streaming_compatible from my example in clang/test/Sema/aarch64-fmv-streaming.c should report an error? 

Yes, that's right, that is because for:
```
(A) __attribute__((target_clones("sve", "simd"))) void ok_arm_streaming(void) __arm_streaming {}
(B) __arm_locally_streaming __attribute__((target_version("sme2"))) void ok_arm_streaming(void) {}
```
(A) needs to be called in streaming mode, whereas (B) should be called in non-streaming mode.

I also think that the following case should give a diagnostic:
```
__attribute__((target_clones("sve", "simd"))) void bad_mixed_streaming2(void) {}

// expected-error@+1 {{multiversioned function declaration has a different calling convention}}
__arm_locally_streaming __attribute__((target_version("sme2"))) void bad_mixed_streaming2(void) __arm_streaming {}
```

The check for `__arm_locally_streaming` prevents that at the moment.

sdesmalen-arm replied:
I also think there is no need to disallow cases where one of the functions is `__arm_streaming_compatible`, e.g.

```
__attribute__((target_clones("sve", "simd"))) void n_callee(void) {}
__attribute__((target_version("sme2"))) void n_callee(void) __arm_streaming_compatible {}
__attribute__((target_version("default"))) void n_callee(void) {}

void s_caller(void) __arm_streaming {
  n_callee();
}
```
because all three versions of `n_callee` can be called from non-streaming mode, so from the perspective of the implementation of `s_caller`, it's fine for it to just do:
```
smstop
bl n_callee
smstart
```

labrinea replied:
Thanks for clarifying! Ok, I'll fix the interaction with locally_streaming in a new revision. Your last example cannot be supported for the reason I explain on the commit message.

labrinea committed:
Changes from last revision:

Made __arm_locally_streaming require the same calling convention
as the rest of the callee versions and updated the tests.

sdesmalen-arm commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -11145,7 +11148,22 @@ bool Sema::areMultiversionVariantFunctionsCompatible(
+    if (OldFPT && NewFPT) {
+      unsigned Diff =
+          OldFPT->getAArch64SMEAttributes() ^ NewFPT->getAArch64SMEAttributes();
+      // Streaming versions cannot be mixed with non-streaming versions.
+      if (Diff & FunctionType::SME_PStateSMEnabledMask)
+        ArmStreamingCCMismatched = true;
+      // Streaming-compatible versions cannot be mixed with anything else.
+      if (Diff & FunctionType::SME_PStateSMCompatibleMask)
+        ArmStreamingCCMismatched = true;
```
nit:
```suggestion
       if (Diff & (FunctionType::SME_PStateSMEnabledMask | 
                   FunctionType::SME_PStateSMCompatibleMask))
         ArmStreamingCCMismatched = true;
```

sdesmalen-arm commented on clang/include/clang/AST/ASTContext.h:
```diff
@@ -3183,6 +3183,25 @@ class ASTContext : public RefCountedBase<ASTContext> {
+      return false;
+
+    for (unsigned I = 0; I < FD1->getNumParams(); ++I)
+      if (!hasSameType(FD1->getParamDecl(I)->getOriginalType(),
+                       FD2->getParamDecl(I)->getOriginalType()))
```
Is there a test (missing) for this condition?

labrinea replied:
As I mentioned in the description of https://github.com/llvm/llvm-project/pull/100181/commits/9b24b8df3afd7c8620c4a7b7edc8533b992dd484 my patch replaces `hasSameType` with `areFMVCompatible` inside `ASTContext::forEachMultiversionedFunctionVersion`. So if I remove this check you highlighted I am seeing the following codegen test failures:
```
Clang :: CodeGenCXX/attr-target-clones-aarch64.cpp
Clang :: CodeGenCXX/attr-target-clones.cpp
Clang :: CodeGenCXX/attr-target-mv-overloads.cpp
Clang :: CodeGenCXX/attr-target-version.cpp
```
but nothing in sema. This suggests we are not diagnosing argument type mismatch. We just deassociate the declaration from the rest of the function versions. My change is NFC. To convince yourlsef look inside `Sema::areMultiversionVariantFunctionsCompatible`, we are only diagnosing mismatch on the return type:
```
enum Different {
  CallingConv = 0,
  ReturnType = 1,
  ...
};
```
If we want to diagnose argument type mismatch then it is not NFC and deserves a separate patch with corresponding sema tests.

labrinea committed:
Changes fro last revision:

Combined two separate SME_PState bitmask checks into one as suggested.

mgabka commented on clang/test/Sema/aarch64-fmv-streaming.c:
```diff
@@ -0,0 +1,46 @@
+__attribute__((target_version("default"))) void ok_arm_streaming(void) __arm_streaming {}
+
+__attribute__((target_clones("sve", "simd"))) void ok_arm_streaming_compatible(void) __arm_streaming_compatible {}
+__arm_locally_streaming __attribute__((target_version("sme2"))) void ok_arm_streaming_compatible(void) __arm_streaming_compatible {}
```
it this a correct test case? can function have attached 2 streaming keyword attributes?

sdesmalen-arm replied:
Yes, the streaming interface can be different from the body of the function. `__arm_locally_streaming` is a declaration attribute and applies the 'streaming' property to the definition (body) of the function, even though the interface remains 'streaming compatible'.

labrinea committed:
Changes from last revision:

* Replaced areFMVCompatible with hasSameType. Looks like this change was
  unnecessary in the first place. Most likely a residue from my WIP
  before I raised the PR. Thanks Sander for finding this!

* Removed the corresponding sema test for variadic type mismatch.

sdesmalen-arm reviewed (approved):
Thanks for addressing the feedback @labrinea, LGTM!

labrinea commented:
/cherry-pick [f8ae128](https://github.com/llvm/llvm-project/commit/f8ae128755777424cf4133e4e8e819c0bc08d2b1)

labrinea commented:
/cherry-pick [f8ae128](https://github.com/llvm/llvm-project/commit/f8ae128755777424cf4133e4e8e819c0bc08d2b1)

labrinea commented:
Created a PR for the backport https://github.com/llvm/llvm-project/pull/101007
